Create a process that shall set QtyCount=QtyBooked - Korrektur für negative Mengen

metas-ts opened the pull request.

metas-ts committed:
Allow M_InventoryLine.QtyCount>0 to be *not* enforced under certain conditions

Also Refactor `QtyCount` validation logic and repository handling; move negative quantity check from `MInventoryLine` to interceptor.

teosarca commented on backend/de.metas.business/src/main/java-legacy/org/compiere/model/MInventoryLine.java:
```diff
@@ -189,12 +192,7 @@ protected boolean beforeSave(final boolean newRecord)
 		}
-
-		// Enforce QtyCount >= 0 - teo_sarca BF [ 1722982 ]
-		if (getQtyCount().signum() < 0)
```
why not simply enforce QtyCount < 0 but tolerate the case when QtyCount=QtyBooked?

metas-ts replied:
agree

teosarca replied:
what was the reason for moving this logic from here to `enforceQtyCountNotNegative` model interceptor?

teosarca reviewed (dismissed):
check my comments

teosarca commented on backend/de.metas.handlingunits.base/src/main/java/de/metas/handlingunits/inventory/InventoryRepository.java:
```diff
@@ -791,5 +797,35 @@ public Set<I_M_InventoryLine> retrieveAllLinesForHU(final @NonNull IContextAware
+		return THREAD_LOCAL_ALLOW_NEGATIVE_QTYCOUNT.get();
+	}
+
+	public void setQtyCountToQtyBookForInventory(@NonNull final InventoryId inventoryId)
```
wondering why using Inventory domain object is not so appealing?

metas-ts replied:
as talked, we think it's OK *in this case*.

teosarca commented:
@metas-ts  cc @cristinamghita  check my review

metas-ts committed:
Follow Teo's review-advise

- Remove thread-local management for negative QtyCount handling, refactor validation logic in `M_InventoryLine` and `InventoryRepository`
- Also, minor improvements

metas-ts committed:
move MI-method to a better place and fix timing

Move `enforceQtyCountNotNegative` logic to `de.metas.inventory.interceptor.M_InventoryLine` and remove it from `de.metas.handlingunits.inventory.interceptor.M_InventoryLine`.

metas-ts commented:
talked with teo an followed his advices

cristinamghita committed:
Merge branch 'intensive_care_hotfix' into intensive_care_hotfix_M_Inventor_Update_negative_qty_count

mrie04 commented on backend/de.metas.handlingunits.base/src/main/java/de/metas/inventory/process/M_Inventory_Update_QtyCount_to_QtyBook.java:
```diff
@@ -22,22 +22,22 @@
 
+import de.metas.handlingunits.inventory.InventoryRepository;
 import de.metas.i18n.AdMessageKey;
-import de.metas.handlingunits.inventory.IInventoryDAO;
```
I assume the DAO itself is also not used anymore now, pls remove

metas-ts committed:
removing obsolete service
